torch/_inductor/comms.py

```diff
@@ -342,6 +342,148 @@ def reorder_compute_and_comm_for_overlap(
     return order
 
 
+def remove_fsdp2_unsharded_param_graph_input_usage(graph: torch.fx.Graph):
+    """
+    This FX graph pass replaces uses of FSDP2 unsharded params with their corresponding
+    graph intermediates that were fsdp.copy_ into the unsharded params in the original graph.
+
+    NOTE: Can only apply this pass to FSDP2 unsharded params that are resharded (i.e. resized to 0)
+    at the end of graph.
+    """
+    node_list = list(graph.nodes)
+
+    # Find all graph inputs and their resize counts
+    graph_input_to_resized_to_full_node_idxes = defaultdict(list)
+    graph_input_to_resized_to_0_node_idxes = defaultdict(list)
+    for idx, node in enumerate(node_list):
+        if (
+            node.op == "call_function"
+            and node.target == torch.ops.inductor.resize_storage_bytes_.default
+            and node.args[0].op == "placeholder"
```

Comment:
nit: is it worth removing this condition and turning it into a defensive assert, something like:
```
assert node.args[0].op == 'placeholder' and node.args[0].meta['val'].untyped_storage().size() == 0
```
? (we only ever resize up from zero-sized graph inputs)

Reply:
I think here there is no guarantee that those resize nodes operate only on FSDP unsharded params, so might not be able to assert they are all graph inputs or that they only ever resize up from zero-sized.

Reply:
discussed offline, updated to assert placeholder.